### DISCUSSION TITLE
Best practice for multiple entity instances?

### DISCUSSION THREAD

**Discussion by mrlubos** [TanStack/query#1125] [Q&A]
Hey everyone! I'm sure this must've been mentioned somewhere already, but cannot find any related discussions.

It's pretty common in REST APIs to have an endpoint returning Entity A and another endpoint returning an array of entities of type A with some additional information. As an example, imagine you have a specific search result (with keys `text`, `id`) and an array of results for the autocomplete dropdown, but this response also includes information about which filters are applied.

As it is, I use 2 `react-query` cache keys for each one of these calls, let's call them `entity` and `entity-list`. All is good until Entity A becomes stale, e.g. when you update it. As an example, imagine a user suggesting edits to the search result because `text` contained incorrect information. What I'd do right now is update the local cache of Entity A (`entity`) with the returned data from the update call. This works, but only if I ensure my array of results (`entity-list`) gets invalidated.

Here's my question. Invalidating related queries is tedious, error-prone, and wasteful because we already have the fresh data in `entity`. Is there a best practice for handling this issue? In Redux, you'd keep each entity in Redux store accessible by `id`. Then you just need to update that object and all your selectors will continue working with the latest data. What's the `react-query` equivalent of this functionality?

Thanks!

**Comment by TkDodo** [comment]
> Invalidating related queries is tedious, error-prone, and wasteful because we already have the fresh data in entity. Is there a best practice for handling this issue?

You can always manually keep the `entity-list` up-to-date after the mutation of the `entity`, something like:

```
useMutation(
    mutateEntity,
    onSuccess: (newData) => {
        queryCache.setQueryData(previousEntityList =>
            previousEntityList.map(
                entity => entity.id === newData.id ? newData : entity
            )
        )
    }
)
```

**Reply by mrlubos** [reply]
Hi @TkDodo, yes, that's the issue I think about when I say it's error-prone to do this manually. This approach requires developers to remember to update any related queries and if they add a new query containing `entity`, they need to know to check all previous relevant queries for potential updates. It would be great if there was an easy way to mark `entity` itself as stale and that would mark all relevant queries as stale. Or we could have a way to update just the affected instance across all queries without invalidating them, though I am not sure how would that API look like.

**Reply by TkDodo** [reply]
yeah, so putting it manually in the cache is somewhat tedious, and re-fetching is wasteful because we need another request. I don't know of a third way though.

> It would be great if there was an easy way to mark entity itself as stale and that would mark all relevant queries as stale.

But wouldn't that also be wasteful, as in your example, when `entity` becomes stale, `entity-list` would need to be re-queried? If you just don't want the query to be executed _now_, but the `entities-list` just to be marked as stale, you can pass the `refetchActive` option to `queryCache.invalidateQueries('entities-list')`. From the [docs](https://react-query.tanstack.com/docs/api/#querycacheinvalidatequeries):

> If you do not want active queries to refetch, and simply be marked as stale, you can use the refetchActive: false option.

**Reply by mrlubos** [reply]
Correct. However, if I had to choose between 2 instances in `react-query` (1 of them stale) and making extra requests, I'd choose option 2. Ideally, we could solve this without making any additional requests or manually tracking entities, would be curious to hear suggestions. For now, I am only thinking about creating some wrappers around `react-query` to store `entity` instances in a separate query, but that doesn't feel right either.

**Reply by TkDodo** [reply]
>  Ideally, we could solve this without making any additional requests or manually tracking entities, would be curious to hear suggestions.

Since there is no "dependency graph" between queries, I wouldn't know how to make that automatically. If you find yourself having a lot of `entity-list` -> `entity-detail` cases, you could make a custom wrapper that always invalidates or manually updates the list in a more generic way, so that you can use it for `user-list / user-detail` as well as `article-list / article-detail` etc, basically:

```
invalidateQueries(`${entityName}-list`)
```

**Reply by mrlubos** [reply]
Hmm, I think this could work for now, thank you @TkDodo! Do you know if there are any plans to add dependency graph to `react-query`?

**Reply by TkDodo** [reply]
I am not aware, but if you have a concrete proposal of how to make it happen, feel free to open an issue and we can discuss it there. I just think that mutations could also potentially need to invalidate different queries depending on parameters / the result from the backend, so I'm not sure you can always pinpoint "migration X invalidates queries Y and Z" in a generic way so that the library can do it for you, but maybe I'm wrong :)

**Reply by mrlubos** [reply]
You're 100% right. A typical example is creating an item where the GET query returns a list of all instances, possibly sorted or otherwise filtered. I think it would be valuable to add at least an example/best practice to docs once we figure out a way to handle stale data due to these actions because I didn't catch it mentioned anywhere. As I mention in the original post, in Redux I used to do this manually and it's A LOT of boilerplate code, so if we can find a way to do this cleanly in `react-query`, that would save people a lot of time. In my mind that should be possible since `react-query` is a "server-side state" cache. 🤞

**Comment by mrlubos** [comment]
Hi @tannerlinsley, I wonder if you have any insights into this topic? Thanks in advance!

**Comment by criticalx7** [comment]
maybe try some normalize cache solution ?

**Reply by mrlubos** [reply]
Hi @criticalx7, thank you for your reply! Can you please elaborate on your suggestion? I'd like to understand it further

**Comment by sergiodxa** [comment]
What I usually do is that if I have an `entity` and `entity-list` I call the keys `["entity", id]` and `["entity", "list"]`. This way I could do `queryCache.invalidateQueries("entity")` and it will invalidate the list and each entity.

**Reply by mrlubos** [reply]
Hi @sergiodxa, that makes sense, thanks! That should solve part of the issue, ensuring all related queries are correctly invalidated. What do you think about being able to update all related entities locally without making any extra requests at all? What I am imagining is `react-query` would internally hold an object with entities accessible by their unique keys (`id`) and if there's any change to the entity, all relevant queries pointing to that instance would automatically return the latest data as they wouldn't hold duplicate state as they do today.
